File: core/src/com/colorflow/screen/PlayScreen.java
package com.colorflow.screen;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.InputMultiplexer;
import com.badlogic.gdx.Screen;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.utils.viewport.ScreenViewport;
import com.colorflow.MainGame;
import com.colorflow.music.BeatDetector;
import com.colorflow.music.IBeatDetector;
import com.colorflow.play.HUDStage;
import com.colorflow.play.PlayStage;
import com.colorflow.play.Score;
import com.colorflow.utility.Position;

public class PlayScreen implements Screen {

    private MainGame game;
    private OrthographicCamera camera, cameraFlipY;
    private Score score;
    private PlayStage playStage;
    private HUDStage hudStage;
    private IBeatDetector beatDetector;
    private State state;
    private InputMultiplexer multiplexer;

    public PlayScreen(MainGame game) {
        this.game = game;
        this.camera = new OrthographicCamera();
        this.camera.setToOrtho(false, Position.getWidthScreen(), Position.getHeightScreen());
        this.camera.update();
        this.cameraFlipY = new OrthographicCamera();
        this.cameraFlipY.setToOrtho(true, Position.getWidthScreen(), Position.getHeightScreen());
        this.cameraFlipY.update();
        this.playStage = new PlayStage(new ScreenViewport(this.cameraFlipY), this);
        this.hudStage = new HUDStage(new ScreenViewport(this.camera), this);
        this.score = new Score();
        this.score.addObserver(hudStage);
        this.beatDetector = new BeatDetector(game.getMusicManager());
        this.multiplexer = new InputMultiplexer();
    }

    @Override
    public void show() {
        reset();
        Gdx.input.setInputProcessor(multiplexer);
    }

    @Override
    public void render(float delta) {
        Gdx.gl.glClearColor(0.2f, 0.2f, 0.2f, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
        playStage.act(delta);
        hudStage.act(delta);
        playStage.draw();
        hudStage.draw();
    }

    @Override
    public void resize(int width, int height) {
    }

    @Override
    public void pause() {
        if (state == State.PLAY) {
            setState(State.PAUSE);
        }
    }

    @Override
    public void resume() {
    }

    @Override
    public void hide() {
    }

    @Override
    public void dispose() {
        playStage.dispose();
        hudStage.dispose();
    }

    public MainGame getGame() {
        return game;
    }

    public State getState() {
        return state;
    }

    public void setState(State state) {
        switch (state) {
            case PLAY:
                game.getMusicManager().play();
                multiplexer.clear();
                multiplexer.addProcessor(playStage);
                multiplexer.addProcessor(playStage.getRing().getListener());
                multiplexer.addProcessor(hudStage);
                break;
            case PAUSE:
                game.getMusicManager().pause();
                multiplexer.clear();
                multiplexer.addProcessor(hudStage);
                break;
            case OVER:
                game.getMusicManager().stop();
                game.getDataManager().incCoins(score.getCoins());
                if (game.getDataManager().getRecord() < score.getPoints()) {
                    game.getDataManager().setRecord(score.getPoints());
                }
                multiplexer.clear();
                multiplexer.addProcessor(hudStage);
                break;
            default:
                throw new IllegalStateException();
        }
        playStage.setState(state);
        hudStage.setState(state);
        this.state = state;
    }

    public Score getScore() {
        return score;
    }

    public void reset() {
        game.getMusicManager().reset();
        score.reset();
        playStage.reset();
        setState(State.PLAY);
    }

    public enum State {
        PLAY, PAUSE, OVER
    }

}

File: core/src/com/colorflow/screen/PlayScreen.kt
package com.colorflow.screen

import com.badlogic.gdx.Gdx
import com.colorflow.stage.HUDStage
import com.colorflow.stage.PlayStage
import com.colorflow.AssetProvider
import com.colorflow.music.BeatSample
import com.colorflow.music.IEventListener
import com.colorflow.music.Music
import com.colorflow.state.GameState
import com.colorflow.state.ScreenType

class PlayScreen(
        game_state: GameState,
        assets: AssetProvider,
        private val music: Music) : UiScreen<PlayStage>(game_state, assets), IEventListener {

    private val play_stage: PlayStage
    private val hud_stage: HUDStage

    private var prev_paused = false

    init {
        play_stage = PlayStage(viewport, game_state, assets)
        hud_stage = HUDStage(viewport, game_state, assets)
        music.add_listener(play_stage)
        music.add_listener(this)
    }

    override fun render(delta: Float) {
        if(state.current_game!!.gameover) {
            game_over()
            return
        }
        if(state.current_game!!.paused && !prev_paused) {
            game_pause()
        }
        if(!state.current_game!!.paused && (prev_paused || !state.current_game!!.started)) {
            game_play()
        }

        if(!state.current_game!!.paused)
            play_stage.act(delta)
        hud_stage.act(delta)
        play_stage.draw()
        hud_stage.draw()

        prev_paused = state.current_game!!.paused
    }

    protected fun game_over() {
        music.stop()
        Gdx.input.vibrate(50)
        assets.get_sound("backspin").play(0.6f)
        state.set_screen(ScreenType.GAME_OVER)
        state.coins += state.current_game!!.score.coins
        if (state.current_game!!.score.points > state.record)
            state.record = state.current_game!!.score.points
        state.persist()
    }

    override fun on_completition() {
        assets.get_sound("complete").play(1f)
        state.set_screen(ScreenType.GAME_OVER)
        state.coins += state.current_game!!.score.coins
        if (state.current_game!!.score.points > state.record)
            state.record = state.current_game!!.score.points
        state.persist()
    }

    protected fun game_pause() {
        if(state.current_game!!.started)
            music.pause()
        multiplexer.clear()
        multiplexer.addProcessor(hud_stage)
    }

    protected fun game_play() {
        if(!state.current_game!!.started) {
            state.current_game!!.started = true
            assets.get_sound("start").play(1f)
        }
        music.play()
        multiplexer.clear()
        multiplexer.addProcessor(play_stage)
        multiplexer.addProcessor(play_stage.get_ring_listener())
        multiplexer.addProcessor(hud_stage)
    }

    override fun pause() {
        game_pause()
    }

    override fun show() {
        play_stage.reset()
        super.show()
    }

    override fun resume() {
        state.current_game!!.paused = true
    }

    override fun dispose() {
        play_stage.dispose()
        hud_stage.dispose()
    }

    override suspend fun on_beat(sample: BeatSample) {}
}